### DISCUSSION TITLE
mail server per domain or one mail server and multiple domain

### DISCUSSION THREAD

**Discussion by ergineroglu** [postalserver/postal#2186] [Help with using Postal]
Hi,

We have just started to explore Postal and have a design question to implement our business needs.
In our case, we will manage multiple senders from multiple domains, and each domain will be one of our customer.
We will be sending transactional mails for these senders.

So what is the best way to implement this:

1. Multiple mail servers with a separate mail server per domain? 
2. One mail server and multiple domains inside it.

We want to separate but not fully distinct ip addresses per sender but tracking and all other stuff can be done with one global configuration.

**Comment by willpower232** [comment]
If you're happy sharing whatever IP addresses are on the server randomly, then it depends entirely on how you want/need to structure the management of your Postal server (i.e. the UI) and credentials.

If you want to use specific IP addresses per customer then you'll have to look into using IP Pools and work around them.

**Reply by ergineroglu** [reply]
Hi Will and thank you for the fast response.

Maybe I can detail our business case in more detail to go over the possibilities. 
From what I understand we can change between those schemas in any time but it will take more effort to move 50+ domains and configuration after we implemented one of it.

First, a short story about us. We are a SaaS personalisation engine for online retailers and personalising shopping behaviour of the customers including email marketing and transactional emails. Up to now, we integrate with third party email service providers like Mailchimp or Klaviyo to send personalised emails and now we want to implement our own mail server. So we are focusing on sending email and incoming mails are not a current concern. And we will not give access to our customers to Postal Admin panel, but use postal as an outgoing email server via smtp and http api. Domains can send between 1.000 - 1.000.000 mails per day so right configuration in performance wise will be also an important consideration. 

Our initial plan is to dedicate 50+ domains (which are online stores) and each domain has one more sender emails (like marketing@myshop.com or info@myshop.com). We will be using open & click tracking which will be done via one domain (our company domain). So domains will be only sharing open & click tracking and ip pool.

Our initial plan is to use a /27 (32 IPs) IP addresses to plan to divide it into 3 chunks to assign each domain to this chunk. 
In future we can increase IPs and may be number of the blocks as well.

So my initial aim is to separate each domain to its own mail server to keep things tidy. Does configuring "50+ mail servers" will create additional resource requirements and performance penalties than "just using one mail server and 50+ sender domains inside"?

Hope I can clearly define our case and objectives, and appreciate your help to go for the right configuration.

**Reply by ergineroglu** [reply]
Hi @willpower232,

What do you think about the options? We have implemented both cases for a small case but still couldn't figure out pros/cons of both optionns?

**Reply by willpower232** [reply]
> We will be using open & click tracking which will be done via one domain (our company domain)

I'm pretty sure that means you can only have one mail server with all the domains in as sharing the click tracking domain between more than one mail server would mean some links wouldn't be found I imagine but if you aren't really rummaging through messages in Postal then it will probably be fine.

I don't really know that much about IP Pools so I'm not sure how well this will work but as long as your server provider allows you to send emails then sure.

> Does configuring "50+ mail servers" will create additional resource requirements and performance penalties than "just using one mail server and 50+ sender domains inside"?

It shouldn't make much difference although you should check the retention settings as that many domains worth of email in a single mail server is likely to get cleared out pretty quickly by default which may not be what you want if you need to do more debugging over a few days.

**Reply by ergineroglu** [reply]
thank you Will,

We will continue with separate configurations (one mail server per customer domain) and learn on the way. 
In future if we see we don't need separate configurations, I think it will be much easier than separating single configuration (one mail server for all customer domains) to multiple ones.

**Comment by huynhphuong10284** [comment]
Just ask more questions here:

- I've setup an smtp for one domain and it's working well. I'm going to add a new domain and I add a new smtp server for it (which belong to an organization). However, when adding a new IP pool for the new domain - it required "HOSTNAME". You know, we setup Postal on one server. So, is Hostname is the same between 2 domains?
- If we have a Postal and going to handle a list domain, do we need to separate smtp or add domain to the same smtp? I've researched and don't know which one is the best. Please advice

**Reply by willpower232** [reply]
I think Postal is just letting you use a friendly name on the IP Pool, I don't think it has any bearing on how Postal behaves.

I don't know what you mean by list domain but I don't think there is any SMTP separation anywhere.

**Reply by huynhphuong10284** [reply]
<img width="756" alt="Screen Shot 2022-10-24 at 16 24 29" src="https://user-images.githubusercontent.com/27182867/197494099-8f6ac838-ac69-4f31-a8aa-2136f12ffcca.png">
<img width="1173" alt="Screen Shot 2022-10-24 at 16 25 14" src="https://user-images.githubusercontent.com/27182867/197494128-99cc7a49-4582-4cbb-a647-f87087536e7e.png">

I'd like to give more information:
- When you build a smtp, you can add 1 or more domains
- When you add a IP pool, it required you input hostname (screenshot). However, when you setup Postal on a server - it has only a hostname. So, can we use a hostname for all pools?

**Reply by willpower232** [reply]
When you create a "mail server" in Postal, you add domains that you can send from (or receive to). This is mostly to help you avoid your messages being detected as spam and also blocks people from sending from incorrect domains.

I can see that the hostname is required but I can't actually see what its used for so I think it is just for reference. I don't use IP Pools so I don't know fully what happens.

**Reply by kluvi** [reply]
Its because reverse DNS rercords. You have to set correct reverse DNS records for IP addresses, which you are using for sending emails. Postal then uses correct hostname in SMTP communication (receiving server can validate, that IP address is resolved back to hostname).

**Reply by huynhphuong10284** [reply]
Ya, it seems we have to configure the same hostname for all IP pools. Let's me test and give you an update if there's any issue.

**Reply by kluvi** [reply]
Please no... You cant have same hostname for all IP addresses. You should have something similar to this:
1)
- IPv4: `1.2.3.4`
- Hostname: `mailer01.example.org`
- DNS: `mailer01.example.org A 1.2.3.4`
- reverse DNS: `4.3.2.1.in-addr.arpa PTR mailer01.example.org`

2)
- IPv4: `1.2.3.5`
- Hostname: `mailer02.example.org`
- DNS: `mailer02.example.org A 1.2.3.5`
- reverse DNS: `5.3.2.1.in-addr.arpa PTR mailer02.example.org`

So it is not possible to configure "same hostname for all IP pools", because one hostname points only to one IP address and one IP address reverse-points to only one hostname. This absolutely have to be unique.

Those IP addresses and domains have nothing with `dns.smtp_server_hostname` in `postal.yml` or anything else. It is just to "authorize" somehow against other mail servers.

The hostname is visible on receiving server, you can check it int mail headers:
```
Received: from mailer01.example.org (mailer01.example.org [1.2.3.4])
        by mail.whatever.com (13.0.3 build 2 RHEL7 x64) with ESMTP (SSL) id 202110650298187525
        for <recipient@whatever.com>; Tue, 25 Oct 2022 02:58:18 +0200
```

Reverse DNS records are useful to preventing spam. And a lot of mail servers just drops messages when you dont have set it up correctly.

**Reply by huynhphuong10284** [reply]
So why Postal supported many domains why one server has only one hostname?

**Reply by kluvi** [reply]
Hostname is assigned to IP address. You can use many IP addresses for sending emails (IP pools). Our setup looks something like this:
```
smtp.example.org A 1.2.3.4
mailer01.example.org A 1.2.3.4
mailer02.example.org A 1.2.3.5
4.3.2.1.in-addr.arpa PTR mailer01.example.org
5.3.2.1.in-addr.arpa PTR mailer02.example.org
```

Domains `mailerXX.example.org` are configured per-ip-address and is visible to SMTP servers and in email headers.
`smtp.example.org` is used in `postal.yml` -> `dns.smtp_server_hostname` - this domain is used in our applications to comunicate with Postal. This domain is not visible in headers. Its not even required, you can use plain IP address for this, or use anything from `mailerXX.example.org`, it will work the same.

But the domains with reverse-dns records are really must-have (when you care about deliverability of your emails).

Reverse DNS is managed by owner of the IP address. So you must "call someone" to set it up. Or if you have assigned whole subnet from your provider, he can delegate reverse DNS for this subnet to your nameservers. (but its more complicated than it looks).
